crates/controller/src/hud/minimap/fill.rs

```diff
@@ -6,22 +9,56 @@ use de_core::{
 use de_map::size::MapBounds;
 use de_objects::SolidObjects;
 use de_terrain::TerrainCollider;
-use de_types::projection::ToFlat;
+use de_types::{objects::*, projection::ToFlat};
 use parry2d::{
     bounding_volume::Aabb,
     math::Point,
     query::{Ray, RayCast},
 };
 
-use super::draw::DrawingParam;
-use crate::ray::ScreenRay;
-
 const TERRAIN_COLOR: Color = Color::rgb(0.61, 0.46, 0.32);
 const PLAYER_COLOR: Color = Color::rgb(0.1, 0.1, 0.9);
-const ENEMY_COLOR: Color = Color::rgb(0.9, 0.1, 0.1);
+const ENEMY_COLORS: [Color; 4] = [
+    Color::rgb(0.1, 0.9, 0.1),
+    Color::rgb(0.9, 0.1, 0.1),
+    Color::rgb(0.9, 0.9, 0.1),
+    Color::rgb(0.1, 0.9, 0.9),
+];
 const MIN_ENTITY_SIZE: Vec2 = Vec2::splat(0.02);
 const CAMERA_COLOR: Color = Color::rgb(0.9, 0.9, 0.9);
 
+#[derive(Resource, Debug)]
+struct PlayerColors {
```

Comment:
I think that the code would be a bit nicer if this becomes `PlayerColors([Color])` and has methods `get_color(Player, ActiveObjectType)`. `get_color()` would then pull the base player color from the hash map and then use a simple `match` on the object type to derive the object color from the base player color.

The benefit of this approach is that:

* `draw_entities_system()` system would be simpler
* we could reuse the struct elsewhere easier in the future (at no cost of extra complexity)
* `player.to_num()` conversion would happen as late as possible (inside `get_color()` method)

Reply:
> Thanks for the guidance! Forgive my noobness. I'm not sure if you meant an array or a slice of `Color` by `PlayerColors([Color])`. I think for slices I would need to manage lifetimes and I'm not confident on how to do that well yet, so I stuck with an array. If it really needs to be a slice I can do more learning and give that a go.

No need for forgiveness :heart: 

I meant array.

> I also couldn't figure out a way to implement a tuple struct where `get_color()` wouldn't be deriving the unit color on every invocation, and when I do that the game slows down tremendously.

It is a bit surprising that it is slow. You basically just convert player to number, take the base color from the array and then "multiply" RGB values in a match.

> So I refactored the closest to what I think you wanted, though the current player is no longer always blue, instead each player gets the color of their player number. It does see more re-usable now, but with `PlayerColors::new([Color; size])` instead of `PlayerColors([Color; size])`.

Ok. I will do another review.

> If you want the playable player to always be blue still, I can refactor again easily and either add a separate method for `get_playable_color()`, or pass `game` or a playable boolean into the existing `get_color()`

It is not necessary. Just make sure that we do not mention the blue in any documentation.

> If there is a way to have tuple struct like `PlayerColors([Color])` have some internal persistent state to cache the derived colors that I just don't know how to do, let me know or point me to some docs.

Reply:
I tried the tuple struct with your simpler algorithm and timed the `get_color()` call, and it's actually faster than doing the hash lookup, which surprised me. I'm not sure why mine was so slow before, perhaps because I was doing a clone on the color?
So it is back to the simplified tuple struct and is not slow.

I wasn't sure how to get the ObjectType coerced to an ActiveObjectType before, but I found a way that works, let me know if there is a more idiomatic way to do it than what I did.

Reply:
Computing the hash and accessing the hash table is actually pretty slow, even though AHashMap is one of the fast ones.